**นายปุณยเกียรติ ถนอมทรัพย์**
**64102010080**

Step 1: โหลดข้อมูล iris ใน scikit 

In [2]:
from sklearn.datasets import load_iris
iris = load_iris()
print(iris)


{'data': array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
     

Step 2: ตรวจสอบว่า ข้อมูลชุดนี้ มีจำนวน feature เท่าใด และ จำนวน record เท่าใด


In [3]:
X = iris['data']
Y = iris['target']

In [4]:
X.shape

(150, 4)

In [5]:
# จำนวน feature ของข้อมูล และ จำนวน sample ของข้อมูล
num_features = X.shape[1]
num_samples = X.shape[0]
print(num_features)
print(num_samples)


4
150


#Question 1: 
จากข้อมูล features ในตัวแปร x และ class label, ให้ตอบคำถามต่อไปนี้


*   จำนวน features = **4**
*   จำนวน samples = **150**



Step 3: แบ่ง train test dataset ในอัตราส่วน train:test = 80:20</br>


In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, stratify=Y)

print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)


(112, 4)
(38, 4)
(112,)
(38,)


#Question 2: 
จากข้อมูลในตัวแปร X_train, X_test, Y_train, Y_test, ให้ตอบคำถามต่อไปนี้
*   จำนวน samples ใน X_train และ Y_train = **112 ทั้งคู่**
*   จำนวน features ใน X_train และ Y_train = **X_train = 4 และ Y_train ไม่มี features เพราะเป็น 1D array**
*   จำนวน samples ใน X_test และ Y_test = **38 ทั้งคู่**
*   จำนวน features ใน X_train และ Y_train = **X_test = 4 และ Y_test ไม่มี features เพราะเป็น 1D array**


Step 4: สร้าง Linear LogisticRegression model โดยอาศัย features ทั้งหมด

In [7]:
logreg_ovo = LogisticRegression(multi_class="multinomial", max_iter=120)# One vs One
logreg_ovr = LogisticRegression(multi_class="ovr", max_iter=120)# One vs Rest

logreg_ovo.fit(X_train[:, 0:4], Y_train)

C:\Users\ireen\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=120, multi_class='multinomial')

Step 5: คำนวณค่า log_loss ของ training set และ test set สำหรับโมเดล

In [8]:
predict_prob = logreg_ovo.predict_proba(X_test[:, 0:4])

In [9]:
from sklearn.metrics import log_loss
log_loss(Y_test,predict_prob)

0.12412150661753514

Step 6: คำนวณค่า mean accuracy  ของ training set และ test set สำหรับโมเดล
REF: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

In [10]:
logreg_ovo.score(X_test, Y_test)

0.9736842105263158

#Question 3: 
* สร้าง Linear LogisticRegression model แบบ OVR และ multimodal โดยอาศัย 2 features (12 โมเดล), 3 features (6 โมเดล) 
* คำนวณค่า log_loss และ ค่า mean accuracy ของ training set และ test set สำหรับแต่ละโมเดล  โดยเปรียบเทียบกับโมเดลที่อาศัย features ทั้งหมด


In [11]:
# Define models for 2 and 3 features
models_2_features = {
    'logreg_ovo2': LogisticRegression(multi_class='multinomial', max_iter=120),
    'logreg_ovr2': LogisticRegression(multi_class='ovr', max_iter=120)
}

models_3_features = {
    'logreg_ovo3': LogisticRegression(multi_class='multinomial', max_iter=120),
    'logreg_ovr3': LogisticRegression(multi_class='ovr', max_iter=120)
}

# Function to calculate log loss_and mean accuracy
def calculate_metrics(model, X_train, X_test, Y_train, Y_test):
    model.fit(X_train, Y_train)
    predict_prob = model.predict_proba(X_test)
    loss = log_loss(Y_test, predict_prob)
    accuracy = model.score(X_test, Y_test)
    return loss, accuracy

# Features definition
features = {
    '2 features': (X_train[:, :2], X_test[:, :2]),
    '3 features': (X_train[:, :3], X_test[:, :3])
}

# Calculate metrics for each model
for feature_name, (X_train_features, X_test_features) in features.items():
    print(f"\nMetrics for {feature_name}:")
    models = models_2_features if '2' in feature_name else models_3_features
    
    for model_name, model in models.items():
        # Calculate metrics
        loss, accuracy = calculate_metrics(model, X_train_features, X_test_features, Y_train, Y_test)
        
        # Print metrics
        print(f'{model_name} - log_loss: {loss}, mean accuracy: {accuracy}')



Metrics for 2 features:
logreg_ovo2 - log_loss: 0.466927649129704, mean accuracy: 0.8421052631578947
logreg_ovr2 - log_loss: 0.538904011547832, mean accuracy: 0.8421052631578947

Metrics for 3 features:
logreg_ovo3 - log_loss: 0.14992436001172998, mean accuracy: 0.9736842105263158
logreg_ovr3 - log_loss: 0.333819426353652, mean accuracy: 0.8947368421052632
